File: TemplateMethodPattern/src/com/company/keliones/KelioneEkonomiskuAutomobiliu.java
package com.company.keliones;

import com.company.kainosSkaiciavimas.Ikainiai;
import com.company.kainosSkaiciavimas.IkainiaiBuilder;

public abstract class KelioneEkonomiskuAutomobiliu extends Kelione {
    private final String transportoPriemonesPavadinimas = "Honda Civic 5 gen.";
    private final Ikainiai ikainiai = new IkainiaiBuilder()
            .setPradineKaina(100)
            .setKilometroKaina(30)
            .setMinutesKaina(14)
            .setVirsytoLaikoKaina(6)
            .setVirsytoAtstumoKaina(5)
            .setParosKaina(1500)
            .setSavaitesKaina(7500)
            .setMenesioKaina(30000)
            .createIkainiai();

    public KelioneEkonomiskuAutomobiliu(int priemoneId, String isvykimoTaskas) {
        super(priemoneId, isvykimoTaskas);
    }

    @Override
    public Ikainiai gautiTransportoPriemonesIkainius() {
        return ikainiai;
    }

    @Override
    public String gautiTransportoPriemonesPavadinima() {
        return transportoPriemonesPavadinimas;
    }

    @Override
    public void paliktiTransportoPriemone() {
        //palikti priemoneId
        System.out.println("Paliktas Honda Civic, id: " + getPriemoneId());
    }

    @Override
    public void pranestiApieNetiketuma() {
        System.out.println("Dekojame uz pranesta Honda Civic, id: " + getPriemoneId() + " apgadinima.");
    }
}

File: TemplateMethodPattern/src/com/company/keliones/KelioneEkonomiskuAutomobiliu.scala
package com.company.keliones

import com.company.kainosSkaiciavimas.Ikainiai

@throws[UnsupportedOperationException]
abstract class KelioneEkonomiskuAutomobiliu(override val priemoneId: Int, override val isvykimoTaskas: String) extends Kelione(priemoneId, isvykimoTaskas) {
  transportoPriemonesPavadinimas = pasirinktiTransportoPriemone(priemoneId)
  if (transportoPriemonesPavadinimas.isEmpty) throw new UnsupportedOperationException("Nepavyko pasirinkti transporto priemones.")
  ikainiai = gautiTransportoPriemonesIkainius

  override def gautiTransportoPriemonesIkainius = new Ikainiai(100, 30, 14, 6, 5, 1522, 7500, 30000)

  override def pasirinktiTransportoPriemone(priemoneId: Int): String = { //if(priemoneID nepasirinktas) pasirinkti
    "Honda Civic 5 gen."
    //else return null;
  }

  override def paliktiTransportoPriemone(): Unit = { //palikti priemoneId
    System.out.println("Paliktas Honda Civic, id: " + priemoneId)
  }

  override def pranestiApieNetiketuma(): Unit = System.out.println("Dekojame uz pranesta Honda Civic, id: " + priemoneId + " apgadinima.")
}